### PR TITLE
Refactor ExprNodes for Python 3 compatibility

### PR DESCRIPTION
**Pull request by EwoutH**
This pull request contains a series of commits refactoring the ExprNodes module to clean up old Python 2 code and be compatible with Python 3.

Most notable changes include:
1. Removal of Python 2 specific elements across various functions such as `_analyse_name_as_type`, `generate_evaluation_code`, and `analyse_as_type`. This involves eliminating outdated type handling, surrogate pair processing, and compatibility checks.
2. Simplification of string handling in alignment with Python 3's consistent Unicode approach, notably in the `StringNode` class and `generate_evaluation_code` method. This includes removing the `unicode_value` attribute and streamlining the handling of string constants.
3. Streamlining of the `ImportNode` to adhere to Python 3 standards, removing Python 2 specific features like `absolute_import` checks and adapting import level handling.

See the individual commit messages for more details.

There are probably many more old Python 2 constructs that can be cleaned up in ExprNodes, but this is a start.

### CODE REVIEW

**Comment by da-woods** [comment]
Probably worth checking that some of this isn't covered by https://github.com/cython/cython/pull/5827 and https://github.com/cython/cython/pull/5830

**Comment by scoder** [comment]
> Probably worth checking that some of this isn't covered by #5827 and #5830

Yeah, I'll merge those first because they also fix some tests and establish behavioural differences.

File: Cython/Compiler/ExprNodes.py [hunk 1]
```diff
@@ -1791,25 +1768,21 @@ def compile_time_value(self, env):
...
     type = str_type
     is_string_literal = True
     is_identifier = None
-    unicode_value = None
 
     def calculate_constant_result(self):
-        if self.unicode_value is not None:
-            # only the Unicode value is portable across Py2/3
-            self.constant_result = self.unicode_value
+        # The value is already in the correct format (Unicode).
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:1781]: This seems a useless comment given the code.
```suggestion
```

File: Cython/Compiler/ExprNodes.py [hunk 2]
```diff
@@ -2124,10 +2097,13 @@ def analyse_as_module(self, env):
 
     def analyse_as_type(self, env):
         type = None
+        # Determine the type based on attributes or context.
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:2100]: This comment seems misleading.
```suggestion
```

File: Cython/Compiler/ExprNodes.py [hunk 3]
```diff
@@ -1791,25 +1768,21 @@ def compile_time_value(self, env):
 
 
 class StringNode(PyConstNode):
-    # A Python str object, i.e. a byte string in Python 2.x and a
-    # unicode string in Python 3.x
+    # In Python 3.x, all strings are Unicode.
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:1771]: I can't really say right now if there is still a use case for this class, but it feels like it still makes a difference with `language_level=2`. Need to look up the string semantics there. I'd rather keep the class as it is right now.
```suggestion
    # An unprefixed string literal.
```

File: Cython/Compiler/ExprNodes.py [hunk 4]
```diff
@@ -2124,10 +2097,13 @@ def analyse_as_module(self, env):
 
     def analyse_as_type(self, env):
         type = None
+        # Determine the type based on attributes or context.
         if self.cython_attribute:
             type = PyrexTypes.parse_basic_type(self.cython_attribute)
         elif env.in_c_type_context:
             type = PyrexTypes.parse_basic_type(self.name)
+
+        # If a type has been determined, return it.
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:2106]: This seems obvious.
```suggestion
```

File: Cython/Compiler/ExprNodes.py [hunk 5]
```diff
@@ -2755,25 +2729,24 @@ class ImportNode(ExprNode):
...
 
     def analyse_types(self, env):
-        if self.level is None:
-            # For modules in packages, and without 'absolute_import' enabled, try relative (Py2) import first.
-            if env.global_scope().parent_module and (
-                    env.directives['py2_import'] or
-                    Future.absolute_import not in env.global_scope().context.future_directives):
-                self.level = -1
-            else:
-                self.level = 0
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:2765]: This is still relevant for `language_level=2`.

File: Cython/Compiler/ExprNodes.py [hunk 6]
```diff
@@ -1755,31 +1755,8 @@ def contains_surrogates(self):
...
-                        self.result_code,
-                        data_cname,
-                        data_cname,
-                        const_code.error_goto_if_null(self.result_code, self.pos)))
-                const_code.put_error_if_neg(
-                    self.pos, "__Pyx_PyUnicode_READY(%s)" % self.result_code)
-            else:
-                self.result_code = code.get_py_string_const(self.value)
+            # In Python 3, all strings are Unicode
+            self.result_code = code.get_py_string_const(self.value)
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:1759]: This seems correct, but the comment is useless without the historical background.
```suggestion
            self.result_code = code.get_py_string_const(self.value)
```

**Review by scoder — COMMENTED** [review]
Sorry, while this PR contains some valid parts, most of these changes are not convincing. I must reject it in its current form.

**Comment by EwoutH** [comment]
@scoder Thanks for rebasing and reviewing. I processed your suggestions.

**Comment by scoder** [comment]
We first need to decide what to do about `UnicodeNode` and `StringNode`: https://github.com/cython/cython/issues/5854

**Pull request closed without merging**